auto generate list of KEPs in annual-report/sig_report go template

Priyankasaggu11929 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
- If this is your first contribution, read our Getting Started guide https://github.com/kubernetes/community/blob/master/contributors/guide/README.md
- If you are editing SIG information, please follow these instructions: https://git.k8s.io/community/generator
  You will need to follow these steps:
  1. Edit sigs.yaml with your change 
  2. Generate docs with `make generate`. To build docs for one sig, run `make WHAT=sig-apps generate`
-->

The PR implements auto generating of the list of KEPs in the annual-report/sig_report.tmpl go template file.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of https://github.com/kubernetes/community/issues/6357


---

One of the examples for the filled KEPs section (in case - "sig-node") looks like following:

![Screenshot 2023-01-13 at 2 37 12 PM](https://user-images.githubusercontent.com/30499743/212281709-d558885c-4543-4bbd-89e0-f8432fe821ff.png)

Priyankasaggu11929 commented:
/hold

palnabarun commented on generator/app.go:
```diff
@@ -686,10 +776,16 @@ func writeYaml(data interface{}, path string) error {
 
 func main() {
+	var err error
+	releases, err = getLastThreeK8sReleases()
```
An improvement to this would be if the generator takes the releases as an optional argument and defaults to the last 3 releases.

This way we can control which releases need to be part of the section if needed.

cblecker replied:
Valid improvement, but also can make a TODO

palnabarun commented on generator/app.go:
```diff
@@ -55,13 +60,93 @@ const (
 	regexRawGitHubURL = "https://raw.githubusercontent.com/(?P<org>[^/]+)/(?P<repo>[^/]+)/(?P<branch>[^/]+)/(?P<path>.*)"
 	regexGitHubURL    = "https://github.com/(?P<org>[^/]+)/(?P<repo>[^/]+)/(blob|tree)/(?P<branch>[^/]+)/(?P<path>.*)"
+
+	// For KEPs automation
+	alpha  = "alpha"
+	beta   = "beta"
+	stable = "stable"
```
These values are defined in https://github.com/kubernetes/enhancements/blob/9d63597d2d85a0ffba5bb4cee0172157948168ce/api/proposal.go#L35-L39 as well. Can we please use them as they are the source of truth for the manifest?

Priyankasaggu11929 replied:
Thanks for pointing. Resolved in the refreshed commit. :)

palnabarun commented on generator/app.go:
```diff
@@ -55,13 +60,93 @@ const (
+	releases         = Releases{}
 )
 
+// KEP represents an individual KEP holding its metadata information.
+type KEP struct {
+	Name            string `json:"name"`
+	Title           string `json:"title"`
+	KepNumber       string `json:"kepNumber"`
+	OwningSig       string `json:"owningSig"`
+	Stage           string `json:"stage"`
+	LatestMilestone string `json:"latestMilestone"`
+}
```
Similar to the above, the KEP structure is also present in https://github.com/kubernetes/enhancements/blob/9d63597d2d85a0ffba5bb4cee0172157948168ce/api/proposal.go#L94

Priyankasaggu11929 replied:
+1. Resolved in the refreshed commit.

palnabarun commented on generator/app.go:
```diff
@@ -55,13 +60,93 @@ const (
+}
+
+func fetchAndFilterKEPs(owningSig string, releases Releases) (map[string][]KEP, error) {
+	url, err := url.Parse(kepURL)
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing url: %v", err)
+	}
+
+	req, err := http.NewRequest("GET", url.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating request: %v", err)
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("Error fetching KEPs: %v", err)
+	}
+	defer resp.Body.Close()
```
nit: This will run O(number of groups) times. Although the number is very trivial, I am slightly worried about reaching GitHub rate limits. If we can fetch the KEPs and cache the result in a variable, that could be best.

cblecker replied:
I think this is returning from GCS not GitHub, so maybe less concerned about rate limits. That said, fetching once and caching still seems like a good idea.

Priyankasaggu11929 replied:
implemented in the refreshed commit. Thanks for pointing.

palnabarun commented on generator/app.go:
```diff
@@ -170,6 +255,7 @@ type Group struct {
 	Meetings         []Meeting
 	Contact          Contact
 	Subprojects      []Subproject `yaml:",omitempty"`
+	KEPs             map[string][]KEP
```
In order to not populate the `sigs.yaml` with an empty dictionary, you can do this.
```suggestion
	KEPs             map[string][]KEP  `yaml:",omitempty"`
```

Priyankasaggu11929 replied:
Thanks for pointing. Resolved in the refreshed commit.

cblecker commented on generator/app.go:
```diff
@@ -55,13 +60,93 @@ const (
+
+	kepsByStage := make(map[string][]KEP)
+	for _, kep := range keps {
+		if kep.OwningSig == owningSig && (kep.Stage == "alpha" || kep.Stage == "beta" || kep.Stage == "stable") && (strings.HasSuffix(kep.LatestMilestone, releases.Latest) || strings.HasSuffix(kep.LatestMilestone, releases.LatestMinusOne) || strings.HasSuffix(kep.LatestMilestone, releases.LatestMinusTwo)) {
```
As @palnabarun highlighted, we can import and use k8s.io/enhancements/api and then use `ValidStages` to loop over things to clean this up a bit

Priyankasaggu11929 commented:
@palnabarun @cblecker, I have addressed all of the review comments mentioned.

I have marked suggested improvements in the `getLastThreeK8sReleases()` function as a TODO and will work on it. 

Thank you. PTAL again.

cblecker commented on generator/annual-report/sig_report.tmpl:
```diff
@@ -14,25 +14,21 @@
-
+{{$releases := getReleases}}
+3. KEP work in {{lastYear}} (v{{$releases.Latest}}, v{{$releases.LatestMinusOne}}, v{{$releases.LatestMinusTwo}}):
 <!--
 In future, this will be generated from kubernetes/enhancements kep.yaml files
 1. with SIG as owning-sig or in participating-sigs
 2. listing 1.x, 1.y, or 1.z in milestones or in latest-milestone
 -->
```
```suggestion
```

We can delete this comment as you've implemented it!

cblecker commented on generator/app.go:
```diff
@@ -686,10 +785,22 @@ func writeYaml(data interface{}, path string) error {
+	// Fetch KEPs and cache them in the keps variable
+	err := fetchKEPs()
+	if err != nil {
+		os.Exit(1)
```
```suggestion
		log.Fatal(err)
```
Can we log these out as fatal for diagnosis?

cblecker commented on generator/app.go:
```diff
@@ -686,10 +785,22 @@ func writeYaml(data interface{}, path string) error {
+
+	releases, err = getLastThreeK8sReleases()
+	if err != nil {
+		os.Exit(1)
```
```suggestion
		log.Fatal(err)
```
Same here

cblecker reviewed (commented):
Tested and works amazing! Couple last comments and this one is good to merge.

Priyankasaggu11929 commented:
Thanks for the review, @cblecker. I've addressed all above comments as well. Just need an lgtm & approve for merge now. Thank you!


/unhold

cblecker reviewed (approved):
/lgtm
